storage/client_range_test.go

```diff
@@ -150,3 +154,158 @@ func TestRejectFutureCommand(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	manualClock := hlc.NewManualClock(0)
+	clock := hlc.NewClock(manualClock.UnixNano)
+	store, stopper := createTestStoreWithEngine(t,
+		engine.NewInMem(proto.Attributes{}, 10<<20),
+		clock,
+		true,
+		nil)
+	defer stopper.Stop()
+
+	// Put an initial value.
+	key := "key"
+	initVal := []byte("initVal")
+	err := store.DB().Put(key, initVal)
+	if err != nil {
+		t.Fatalf("failed to put: %s", err)
+	}
+
+	waitPut := make(chan struct{})
+	waitFirstGet := make(chan struct{})
+	waitTxnRestart := make(chan struct{})
+	waitSecondGet := make(chan struct{})
+	waitTxnComplete := make(chan struct{})
+
+	// Start a writer.
+	go func() {
+		epoch := -1
+		// Start a txn that does read-after-write.
+		// The txn will be restarted twice, and the out-of-order put
+		// will happen in the second epoch.
+		if err := store.DB().Txn(func(txn *client.Txn) error {
+			epoch++
+
+			if epoch == 1 {
+				// Wait until the second get operation is issued.
+				close(waitTxnRestart)
+				<-waitSecondGet
+			}
+
+			updatedVal := []byte("updatedVal")
+			if err := txn.Put(key, updatedVal); err != nil {
+				return err
+			}
+
+			actual, err := txn.Get(key)
+			if err != nil {
+				return err
+			}
+			if !bytes.Equal(actual.ValueBytes(), updatedVal) {
+				t.Fatalf("unexpected get result: %s", actual)
+			}
+
+			if epoch == 0 {
+				// Wait until the first get operation will push the txn timestamp.
+				close(waitPut)
+				<-waitFirstGet
+			}
+
+			b := &client.Batch{}
+			err = txn.Commit(b)
+			return err
+		}); err != nil {
+			t.Fatal(err)
+		}
+
+		if epoch != 2 {
+			t.Fatalf("unexpected number of txn retries: %d", epoch)
+		}
+
+		close(waitTxnComplete)
+	}()
+
+	<-waitPut
+
+	// Advance the clock and send a get operation with higher
+	// priority to trigger the txn restart.
+	manualClock.Increment(100)
+
+	priority := int32(math.MaxInt32)
+	requestHeader := proto.RequestHeader{
+		Key:          proto.Key(key),
+		RaftID:       1,
+		Replica:      proto.Replica{StoreID: store.StoreID()},
+		UserPriority: &priority,
+		Timestamp:    clock.Now(),
+	}
+	getCall := proto.Call{
+		Args: &proto.GetRequest{
+			RequestHeader: requestHeader,
+		},
+		Reply: &proto.GetResponse{},
+	}
+	err = store.ExecuteCmd(context.Background(), getCall)
+	if err != nil {
+		t.Fatalf("failed to get: %s", err)
+	}
+
+	// Wait until the writer restarts the txn.
+	close(waitFirstGet)
+	<-waitTxnRestart
+
+	// Advance the clock and send a get operation again. This time
+	// we use TestingCommandFilter so that a get operation is not
+	// processed after the write intent is resolved (to prevent the
+	// timestamp cache from being updated).
+	manualClock.Increment(100)
+
+	requestHeader.Timestamp = clock.Now()
+	getCall = proto.Call{
+		Args: &proto.GetRequest{
+			RequestHeader: requestHeader,
+		},
+		Reply: &proto.GetResponse{},
+	}
+
+	numGets := 0
+	storage.TestingCommandFilter = func(args proto.Request, reply proto.Response) bool {
```

Comment:
@kkaneda as mentioned in the other thread, setting `storage.TestingCommandFilter` during a test is a bad idea. Can you set it once at the top and using `atomic` to manipulate `numGets` when you need to?